server/src/Main.hs

```diff
@@ -143,27 +199,118 @@ app origin rpIdHash db pending registryVar = do
       Just name -> Scotty.json $ String $ WA.unUserAccountName name
   Scotty.get "/logout" $ logout db
 
-mkCredentialDescriptor :: WA.CredentialEntry -> WA.CredentialDescriptor
-mkCredentialDescriptor WA.CredentialEntry {WA.ceCredentialId, WA.ceTransports} =
-  WA.CredentialDescriptor
-    { WA.cdTyp = WA.CredentialTypePublicKey,
-      WA.cdId = ceCredentialId,
-      WA.cdTransports = Just ceTransports
-    }
+-- | In this function we receive the intent of the client to register and reply
+-- with the
+-- [creation options](https://www.w3.org/TR/webauthn-2/#dictionary-makecredentialoptions)
+-- . This function also checks if the specified user hasn't already registered.
+-- WebAuthn does allow a single user to register multiple credentials, but this
+-- server doesn't implement it.
+beginRegistration :: Database.Connection -> PendingCeremonies -> Scotty.ActionM ()
+beginRegistration db pending = do
+  req@RegisterBeginReq {accountName, accountDisplayName} <- Scotty.jsonData @RegisterBeginReq
+  Scotty.liftAndCatchIO $ TIO.putStrLn $ "Register begin <= " <> jsonText req
+  exists <- Scotty.liftAndCatchIO $
+    Database.withTransaction db $ \tx -> do
+      Database.userExists tx (WA.UserAccountName accountName)
+  when exists $ Scotty.raiseStatus HTTP.status409 "Account name already taken"
+  userId <- Scotty.liftAndCatchIO WA.generateUserHandle
+  let user =
+        WA.CredentialUserEntity
+          { WA.cueId = userId,
+            WA.cueDisplayName = WA.UserAccountDisplayName accountDisplayName,
+            WA.cueName = WA.UserAccountName accountName
+          }
+  options <- Scotty.liftAndCatchIO $ insertPendingCeremony pending $ defaultPkcco user
+  Scotty.liftAndCatchIO $ TIO.putStrLn $ "Register begin => " <> jsonText options
+  Scotty.json $ WA.encodeCredentialOptionsRegistration options
+
+-- | Completes the registration procedure. Receives the credential from the
+-- client and performs the registration ceremony. If the ceremony succeeds, the
```

Comment:
Terminology is a bit off. It's the registration ceremony that is being completed by performing the [relying party operation](https://www.w3.org/TR/webauthn-2/#sctn-rp-operations)